sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala

```diff
@@ -463,6 +437,442 @@ class MultiStatefulOperatorsSuite
     )
   }
 
+  test("stream-stream time interval left outer join -> aggregation, append mode") {
+    val input1 = MemoryStream[(String, Timestamp)]
+    val input2 = MemoryStream[(String, Timestamp)]
+
+    val s1 = input1.toDF()
+      .selectExpr("_1 AS id1", "_2 AS timestamp1")
+      .withWatermark("timestamp1", "0 seconds")
+      .as("s1")
+
+    val s2 = input2.toDF()
+      .selectExpr("_1 AS id2", "_2 AS timestamp2")
+      .withWatermark("timestamp2", "0 seconds")
+      .as("s2")
+
+    val s3 = s1.join(s2, expr("s1.id1 = s2.id2 AND (s1.timestamp1 BETWEEN " +
+      "s2.timestamp2 - INTERVAL 1 hour AND s2.timestamp2 + INTERVAL 1 hour)"), "leftOuter")
+
+    val agg = s3.groupBy(window($"timestamp1", "10 minutes"))
```

Comment:
How is this different from `join() followed deDup()` test in EventTimeWatermarkSuite.scala? The latter was not allowed.

Reply:
Explained in other comment - https://github.com/apache/spark/pull/39931#issuecomment-1441343786

Reply:
`groupBy` here gets two event timestamp columns too, right, similar to dedup(). What is the difference?

Reply:
> For the test case of join + groupBy, we picked one event time column for time window and discarded other one.

Moving the discussion to the comment thread here.
Could you expand the rationale here? What is the watermark here? Who decides which timestamp is picked? What is the watermark based on (or is it independent of event timestamp picked?)

Reply:
May be I should read the comment in the test in full. Do you think that will clarify this?

Reply:
> Many stateful operators including streaming aggregation only takes columns in grouping keys into consideration when figuring out event time column. The test case uses count(*) which feels like all columns are being used, but columns in non-grouping key are not effective on determining event time columns.

Could you point to the code where this decision is made? Trying to see why it matters since OutputWatermark for join() and groupBy() don't change even if we pick one or the other. 

I.e. what would go wrong if we picked the wrong event-time column due to a bug?

Reply:
I've been thinking through this again... I might be probably overlooked.

My thought was that if there are two event time columns, the row should wait for the watermark to pass by two values (W > et1 && W > et2) so that further input won't match with the row. But it seems like the condition is OR, not AND. If either W > et1 or W > et2 happens, further input won't match with the row, as W is the lower bound of both columns.

I probably need to revisit this tomorrow again (ideal), or postpone to the next JIRA ticket. If you can help me finding some test/use case which shows expected behavior, that should be awesome.

Reply:
May be we can leave a comment and open to ticket to address it as a follow up. But I think it is an important issue as it relates to core of the feature. I will think about it deeper to. My question at high level is "what is the difference that makes this ok for windowed count, but not ok for deDup()"

Could you point to the code where this difference comes from? I will look deeper as well.

Reply:
dedup() with dropping one of event time column from the list of columns should also work. That's not about which operator it is. It's just that a stateful operator can receive two event time columns, and current logic does not consider the case (or we did it but I don't get it).

Let me think about this once more. We probably wouldn't like to disallow the functionality first and allow it later, hence it sounds to me we need to give a try to avoid disallowing the functionality.

Reply:
https://github.com/apache/spark/pull/39931/files#diff-da6ad0bc819dce994a16436fa0797bfc8484644b475227a04c2a3eb5927515f7R352-R386

^^^ The existing logic only handles one event time column. We create an expression based on figured out column in caller site.

Reply:
I see more and more places which simply pick up the first occurrence of event time column. (Logic in stream-stream join also assumes there is only one event time column which I missed. Will add assertion.)

That said, considering multiple event time columns for all stateful operators needs non-trivial reasoning. Let's leave this as a follow-up.

Reply:
https://issues.apache.org/jira/browse/SPARK-42549